File: STA5106_Comp_Statistics/HW10/main_1.py
# Problem 1: Importance sampling to estimate an integral

import numpy as np
import matplotlib.pyplot as plt

np.random.seed(2019)
y = 0.5
N_range = 10 ** np.arange(1, 7, 0.1)
est_seq = [];
for i in range(len(N_range)):
    data = np.random.exponential(1/3, int(N_range[i]))
    Z_est = np.mean(np.exp(-(y - data)**2/2)) / 3
    theta_est = np.mean(data * np.exp(-(y - data)**2/2)) / (3 * Z_est)
    est_seq.append(theta_est)

plt.figure(1)
plt.plot(np.log10(N_range), est_seq, linewidth=2);
plt.xlabel("Sample size ($log_{10}$)")
plt.ylabel("Estimate")
plt.title("Problem 4 - Importance sampling estimate", fontsize=10)
plt.savefig("p1_py.png")

File: STA5106_Comp_Statistics/HW10/main_1.m
% Problem 1: Importance sampling to estimate an integral

clear;

rng(2019);
y = 0.5;
N_range = round(10.^(1:0.1:7));
est_seq = [];
for i = 1:length(N_range)
    data = exprnd(1/3, 1, N_range(i)); % Parameter = 3, Generate (1 x i) samples
    Z_est = mean(exp(-(y - data).^2/2)) / 3;
    theta_est = mean(data .* exp(-(y - data).^2/2)) / (3 * Z_est);
    est_seq(i) = theta_est;
end

figure(1); clf;
plot(log10(N_range), est_seq, 'LineWidth', 2);
xlabel("Sample size (log_{10})");
ylabel("Estimate");
title("Problem 1 - Importance sampling estimate", "FontSize", 14);
